R/set_data.R

```diff
@@ -44,20 +44,66 @@ mash_set_data = function(Bhat,Shat=NULL,alpha=0,df=Inf,pval=NULL,V=diag(ncol(Bha
   } else {
     Shat_alpha = matrix(1, nrow(Shat), ncol(Shat))
   }
-  return(list(Bhat=Bhat, Shat=Shat, Shat_alpha=Shat_alpha, V=V, alpha=alpha))
+  data = list(Bhat=Bhat, Shat=Shat, Shat_alpha=Shat_alpha, V=V, alpha=alpha)
+  class(data) = 'mash'
+  return(data)
 }
 
 #' Deprecated function call of `mash_set_data`.
 #' @export
-
 set_mash_data = function(Bhat,Shat=NULL,alpha=0,df=Inf,pval=NULL,V=diag(ncol(Bhat))){
   .Deprecated("mash_set_data")
   mash_set_data(Bhat,Shat,alpha,df,pval,V)
 }
 
+#' Create a data object for mash contrast analysis
+#' @param mashdata a mash data object containing the Bhat matrix, standard errors, V; created using set_mash_data
+#' @param L the contrast matrix
+#' @return a data object after the contrast transfermation
+#' @export
+mash_set_data_contrast = function(mashdata, L){
+  # check data
+  if(class(mashdata) != 'mash'){
+    stop('data is not a "mash" object')
+  }
+
+  # check contrast
+  R = ncol(mashdata$Bhat)
+  if(ncol(L) != R){
+    stop('The contrast is not correct')
+  }
+
+  # transfer Bhat
+  Bhat = mashdata$Bhat %*% t(L)
+  mashdata$Shat_orig = mashdata$Shat
+  mashdata$L = L
+
+  # get standard error for delta
+  if(is_common_cov_Shat(mashdata)){
+    V = get_cov(mashdata,1) # all covariances are same
+    Shat = matrix(rep(sqrt(diag(V)), each=nrow(Bhat)), nrow = nrow(Bhat))
+  } else{
+    Shat = t(sapply(1:nrow(Bhat), function(j){
+      V = get_cov(mashdata,j)
+      return(sqrt(diag(V)))
+    }))
+  }
+
+  data = list(Bhat = Bhat, Shat=Shat,
+              Shat_orig = mashdata$Shat_orig,
+              Shat_alpha = matrix(1, nrow(Shat), ncol(Shat)),
+              V = mashdata$V, alpha = 0, L = L)
+  return(data)
```

Comment:
don't you want 'class(data) = 'mash'' here?

Reply:
Yes, I check the class on line 66

Reply:
I mean don't you want to set the class of the new data object you have created to be 'mash'.
It seems line 92 creates a new data object, over-writing the old one, so as written this will not return an object of class 'mash'.  Or am i wrong?

Reply:
I see.